### ISSUE TITLE
add fuzzy searching instead of just globbing

### ISSUE THREAD

**Issue by kcolford** [dannyvankooten/browserpass#32]

**Comment by emersion** [comment]
That would be a great improvement. Since we already sort-of list all files in the password store, this shouldn't impact performance too much. I'll look if i can do it.

**Comment by kcolford** [comment]
You could add something in so that if you're on `www.facebook.com` then it will match `facebook.com` (and maybe even all subdomains thereof) but keep it in order of most precise.

And then just simple substring matching for the rest would be useful.

**Comment by emersion** [comment]
I think it already does a simple substring matching, doesn't it?

**Comment by kcolford** [comment]
no, it does a prefix matching (typing in `facebook` brings up `facebook.com` but not `www.facebook.com`)

**Comment by tarciosaraiva** [comment]
+1

**Comment by dannyvankooten** [comment]
I like this for subdomains but am not too sure about enabling this for just _any_ prefix (like it does not with suffixes, as a wildcard). For example, I think when searching for "goo" it should return the following.

```
search: goo
results: 
  - google.com
  - goofoo.com
  - accounts.google.com
```

But not "foogoo.com", as I think some kind of naming consistency / rule is important so the search result list doesn't get cluttered. Do you agree?

**Comment by kcolford** [comment]
I definitely agree, although I think we should distinguish between searches
entered by the user and automatic search by domain name. We can't allow a
similar URL being used as phishing, yet we still want to allow the user
some leaway with typos that they enter.

**Comment by travisghansen** [comment]
User searches also appear to be case sensitive.  Can that be changed?

**Comment by 00dani** [comment]
When you're on a subdomain like accounts.google.com, passwords for google.com should definitely be found in the automatic search. Currently I have to manually type the site's name for many sites, since my passwords often aren't organised into the exact subdomain the login page lives on.

Perhaps if you're on a domain foo.bar.baz.quux, Browserpass should automatically search for passwords matching "foo.bar.baz.quux", "bar.baz.quux", and "baz.quux", with results that match the longer search strings being ranked higher?

**Comment by maximbaz** [comment]
Are you on the latest browserpass version? I have passwords saved in `personal/google.com/username` and they are being *automatically* discovered on `accounts.google.com`. (If you are using the browserpass installed from Webstore, then you are **not** on the latest version - #117).

*Manual* search capability is not as good and we have two feature requests to improve it:
* #49 is about when you have a password stored as `personal/account.google.com/username`, it is not being discovered by manual search for `google`.
* This request is  about when you have a password stored as `personal/account.google.com/username`, it is not being discovered by manual search for `ago` (**a**ccount.**go**gle)

If you are on the latest browserpass version and still experience the issue with *automatic* search, please report a separate issue.

**Comment by 00dani** [comment]
Oh! My mistake, I was indeed on 1.0.5. I've just updated to 1.0.14 and it looks like automatic discovery is working as you describe. Sorry for the confusion. 😳

**Comment by rettgerst** [comment]
I'm in the habit of storing pass entries as _just_ the domain name (without the tld), and if I have multiple accounts for a service, then named by that username in a folder for that service. for example my entry for this account is `github/rettgerst[.gpg]`. it would be great, for my use case, if browserpass could do a substring match of the full relative path based on the domain name (again, minus the tld).

**Comment by maximbaz** [comment]
As I mentioned above, this ticket is about *manual search*, specifically fuzzy search, so when you type `gitret` you find a match for `github/rettgerst`. 

If you are referring to browserpass *automatically* matching website without TLD, then this has been already raised in #183 and I can tell you it's not going to happen - it violates the primary purpose of the browser extension to protect you from phishing attacks. If you need to reuse the same credentials on multiple websites (e.g. `amazon.com` and `amazon.co.uk`), the recommended approach is to use symlinks - that way you deliberately create a whitelist of domains that can access certain credentials.

**Comment by rettgerst** [comment]
ah, as I come from passff which does auto-match domains to entries, I hadn't made the connection that browserpass' lack thereof was a feature, and not just a fault of the matching algorithm. this makes perfect sense.

**Comment by maximbaz** [comment]
Great - and just to make it clear, browserpass does automatch domains to entries, but only when it is 100% sure that this is the correct website and not a phishing one - in other words, if your file/folder name contains a full domain name, e.g. `github.com.gpg` or `github.com/myusername.gpg`.

**Comment by maximbaz** [comment]
@sophacles implemented this in #213, will be released shortly in `2.0.12`.
